mongo/gridfs/gridfs_examples_test.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package gridfs_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo/gridfs"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func ExampleBucket_OpenUploadStream() {
+	var fileContent []byte
+	var bucket *gridfs.Bucket
+
+	// Specify the Metadata option to include a "metadata" field in the files collection document.
+	uploadOpts := options.GridFSUpload().SetMetadata(bson.D{{"metadata tag", "tag"}})
+	uploadStream, err := bucket.OpenUploadStream("filename", uploadOpts)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err = uploadStream.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	// Use SetWriteDeadline to force a timeout if the upload does not succeed in 2 seconds.
+	if err = uploadStream.SetWriteDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+
+	if _, err = uploadStream.Write(fileContent); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExampleBucket_UploadFromStream() {
+	var fileContent []byte
+	var bucket *gridfs.Bucket
+
+	// Use SetWriteDeadline to force a timeout if the upload does not succeed in 2 seconds.
+	if err := bucket.SetWriteDeadline(time.Now().Add(2 * time.Second)); err != nil {
+		log.Fatal(err)
+	}
+	// To make sure the deadline only applies to this upload, set the deadline to an empty time.Time{} after the
```

Comment:
This is rather odd.  Does it mean you can't have concurrent uploads on the same bucket with different write deadlines?

Reply:
Yes. Unfortunately, this was the best way to do it given that we wanted our types to implement the `io.Reader`/`io.Writer` interfaces and wanted consistency between `Bucket` and `UploadStream`/`DownloadStream`. If Go ever exposes new interfaces that take contexts, we can likely deprecate a lot of this code, but we're stuck with this for now.

Reply:
I think that we should not show this usage pattern, as it seems unsafe for concurrent usage.  I can see setting the read and write deadline for the Bucket once on creation, but not to attempt to modify it per upload/download and then unset it afterwards. 

If an application requires per-operation deadlines, they should either create a new Bucket, or use the variants that return UploadStream/DownloadStream.